source/math/FloatConvexHull2D.ooc

```diff
@@ -21,55 +21,126 @@ import FloatTransform2D
-		// TODO: Re-compute convex hull from new points if necessary, i.e. if transform changes order of points
 		newPoints := VectorList<FloatPoint2D> new(this count)
 		for (i in 0 .. this count)
 			newPoints add(transform * this _points[i])
 		This new(newPoints)
 	}
-	_isOnLeft: static func (leftPoint, rightPoint, queryPoint: FloatPoint2D) -> Bool {
-		(rightPoint x - leftPoint x) * (queryPoint y - leftPoint y) > (rightPoint y - leftPoint y) * (queryPoint x - leftPoint x)
+	computeHull: func {
+		if (this _hull)
+			this _hull clear()
+		else
+			this _hull = VectorList<FloatPoint2D> new()
+		
+		if (this _points count <= 2)
+			this _hull = this _points copy()
+		else {
+			// Uses the Quickhull algorithm if more than two points, average complexity O(n log n)
+			// http://www.cse.yorku.ca/~aaw/Hang/quick_hull/Algorithm.html
+			leftMostIndex := 0
+			rightMostIndex := 0
+			for (i in 0 .. this _points count)
+				if (this _points[i] x < this _points[leftMostIndex] x)
+					leftMostIndex = i
+				else if (this _points[i] x > this _points[rightMostIndex] x)
+					rightMostIndex = i
+			leftEndpoint := this _points[leftMostIndex]
+			rightEndpoint := this _points[rightMostIndex]
+			leftSet := VectorList<FloatPoint2D> new()
+			rightSet := VectorList<FloatPoint2D> new()
+			for (i in 0 .. this _points count)
+				if (i != leftMostIndex && i != rightMostIndex)
+					if (This _isOnLeft(this _points[leftMostIndex], this _points[rightMostIndex], this _points[i]))
+						leftSet add(this _points[i])
+					else
+						rightSet add(this _points[i])
+			this _hull add(leftEndpoint)
+			this _findHull(leftSet, leftEndpoint, rightEndpoint)
+			this _hull add(rightEndpoint)
+			this _findHull(rightSet, rightEndpoint, leftEndpoint)
+		}
 	}
-	free: override func {
-		this _points free()
-		super()
+	_findHull: func (currentSet: VectorList<FloatPoint2D>, leftPoint, rightPoint: FloatPoint2D) {
+		if (currentSet count == 1)
+			this _hull add(currentSet[0])
+		else if (currentSet count > 1) {
+			maximumDistance := Float negativeInfinity
+			maximumDistanceIndex := -1
+			for (i in 0 .. currentSet count) {
+				distance := This _hullLinePseudoDistance(leftPoint, rightPoint, currentSet[i])
+				if (distance > maximumDistance) {
+					maximumDistance = distance
+					maximumDistanceIndex = i
+				}
+			}
+			maximumDistancePoint := currentSet[maximumDistanceIndex]
+			outsideLeft := VectorList<FloatPoint2D> new()
+			outsideRight := VectorList<FloatPoint2D> new()
+			for (i in 0 .. currentSet count)
+				if (i != maximumDistanceIndex)
+					if (This _isOnLeft(leftPoint, maximumDistancePoint, currentSet[i]))
+						outsideLeft add(currentSet[i])
+					else if (This _isOnLeft(maximumDistancePoint, rightPoint, currentSet[i]))
+						outsideRight add(currentSet[i])
+			this _findHull(outsideLeft, leftPoint, maximumDistancePoint)
+			this _hull add(maximumDistancePoint)
+			this _findHull(outsideRight, maximumDistancePoint, rightPoint)
+		}
+		currentSet free()
```

Comment:
As a rule, I would strongly advise you not to free input variables here, but in the function where they were created.